# CSE 5243 - Introduction to Data Mining
## Homework 2: Classification
- Semester: Fall 2022
- Instructor: Greg Ryslik
- Section: Wednesday 12:45
- Student Name: John Smith
- Name.#: smith.2@osu.edu
***

# Introduction

### Objectives

In this lab, you will use the "Cleveland Data CLEANED AND TRIMMED.csv" heart disease dataset provided on Carmen.  It is a subset of the "Cleveland" dataset that can be found here: https://archive.ics.uci.edu/ml/datasets/Heart+Disease

You will configure, execute, and evaluate an off-the-shelf K-Nearest-Neighbor classifier and two other classifiers you choose.

The objectives of this assignment are:
1.	Understand how to evaluate classifiers based on business criteria.
2.	Understand how to tune and evaluate a classifier to achieve good performance.
3.	Understand how to select and evaluate suitable off-the-shelf classifiers based on the characteristics of a dataset and the outcomes you need.

### Problem Statement

You work for a medical institution that wants to improve the heart health of its patients.  You have obtained a dataset that contains a variety of demographic and health-related information for a group of patients.  It also includes a CLASS variable "num" that indicates the heart health of each of the patients.  The values are:

0. No heart health issues / risk
1. Slight heart health issues / risk
2. Moderate heart health issues / risk
3. High heart health issues / risk
4. Extreme heart health issue / risk

You have been asked to develop a classifier based on the dataset data, to predict the CLASS of new patients so they can be enrolled in interventions based on their demographic data.

The COSTs of the interventions are as follows, based on the predicted class of each patient

0) Tiny intervention: 100 (dollars)
1) Minor intervention: 200
2) Moderate intervention: 300
3) Significant intervention: 400
4) Extreme intervention: 500

The BENEFITs of the interventions are as follows:

- If the classification is correct and the correct intervention given to the patient: 500 * (TRUE_CLASS + 1)
- If an incorrect classification is made and the wrong intervention is given to the patient: 0

You would like to find a classifier that maximizes the overall **NET_BENEFIT = BENEFIT - COST**. Therefore, a larger positive number is a good outcome.

So, for example:

- If a patient's true class is 4 and their predicted class is 0, their NET_BENEFIT = 0 - 100 = -100
- If a patient's true class is 4 and they are correctly classified, their NET_BENEFIT is 500*(4+1) - 500 = 2000

The medical institution would like you to evaluate the use of a K-Nearest-Neighbor classifier as a starting point.  You agree to do so, as long as you then can choose a different classifier if you are not satisfied with KNN.

### Collaboration
For this assignment, you should work as an individual. You may informally discuss ideas with classmates, but your work should be your own.

### What you need to turn in:
1)	Code

-	For this homework, the code is the Jupyter Notebook.  Use the provided Jupyter Notebook template, and fill in the appropriate information.
-	You may use common Python libraries for I/O, data manipulation, data visualization, etc. (e.g., NumPy, Pandas, MatPlotLib,…  See reference below.) 
-	You may not use library operations that perform, in effect, the “core” computations for this homework (e.g., If the assignment is to write a K-Means algorithm, you may not use a library operation that, in effect, does the core work needed to implement a K-Means algorithm.).  When in doubt, ask the grader or instructor.
-	The code must be written by you, and any significant code snips you found on the Internet and used to understand how to do your coding for the core functionality must be attributed.  (You do not need to attribute basic functionality – matrix operations, IO, etc.)
-	The code must be commented sufficiently to allow a reader to understand the algorithm without reading the actual Python, step by step.
-	When in doubt, ask the grader or instructor.

2)	Written Report
-	For this homework, the report is the Jupyter Notebook.  The report should be well-written.  Please proof-read and remove spelling and grammar errors and typos.
-	The report should discuss your analysis and observations. Key points and findings must be written in a style suitable for consumption by non-experts.  Present charts and graphs to support your observations. If you performed any data processing, cleaning, etc., please discuss it within the report.

### Grading

1.	Overall readability and organization of your report (5%)
> - Is it well organized and does the presentation flow in a logical manner?
> - Are there no grammar and spelling mistakes?
> - Do the charts/graphs relate to the text?
> - Are the summarized key points and findings understandable by non-experts?
> - Do the Overview and Conclusions provide context for the entire exercise?
2.	Evaluation Method (10%)
> - Does your evaluation method meet the needs of the developer (you) as well as the needs of your business stakeholders?
> - Is the evaluation method sound?
> - Did you describe both the method itself and why you chose it?
3.	Pre-Processing of the Dataset (10%)
> - Did you make reasonable choices for pre-processing, and explain why you made them?
4.	Evaluation of the KNN Classifier (20%)
> - Is your algorithm design and coding correct?
> - Is it well documented?
> - Have you made an effort to tune it for good performance?
> - Is the evaluation sound?
5.	Evaluation of the Second Classifier (20%)
> - Is your algorithm design and coding correct?
> - Is it well documented?
> - Have you made an effort to tune it for good performance?
> - Is the evaluation sound?
6.	Evaluation of the Third Classifier (20%)
> - Is your algorithm design and coding correct?
> - Is it well documented?
> - Have you made an effort to tune it for good performance?
> - Is the evaluation sound?
7.	Comparison of the Three Classifiers (10%)
> - Is the comparison sound?
> - Did you choose a specific classifier as best and explain why?
8.  Conclusions (5%)
> - Did you summarize appropriately your critical findings. 
> - Did you provide appropriate conclusions and next steps.

### How to turn in your work on Carmen:

Submit to Carmen the Jupyter Notebook, the html print out of your Jupyter notebook, and any supporting files that you used to process and analyze this data. You do not need to include the input data.  All submitted files (code and/or report) except for the data should be archived in a *.zip file, and submitted via Carmen.  Use this naming convention:
 
•	HW2_Surname_DotNumber.zip

The submitted file should be less than 10MB.

***

***
# Section: Overview
- Insert a short description of the scope of this exercise, any supporting information, etc.
***

This data set is about a group of patients with heart disease. The project is to develop a classifier to properly enroll new patients and give advice on the degree of intervention. We want to minimize the cost of intervention but to discover as more patients accurately as possible, where more severe patients are prioritized.

***
# Section: Setup
- Add any needed imports, helper functions, etc., here.
***

In [2]:
from sklearn.model_selection import train_test_split
import pandas as pd

***
# Section: 1 - Evaluation Method
- Define measures for evaluating the classification models you develop.  Explain why the measures you choose provide a useful view into the value and usefulness of the model you eventually chose for the company to use.  Define two types:
***

***
## Section: 1.1 - Define measures that do not include the cost information
- (e.g., confusion matrices, accuracy, precision, recall, F-measures, etc.).
- Consider using: from sklearn.metrics import classification_report, confusion_matrix
***

In [13]:
from sklearn.metrics import classification_report, confusion_matrix

***
## Section: 1.2 - Define measures that do include the cost information
- (e.g., using cost matrices).
- Consider creating a function that takes a confusion matrix and calculates the cost, like this:
***

In [14]:
def calculate_cost(conf_matrix):
# Fill in the cost matrix values
#                           PREDICTED VALUES
#                      0     1     2     3     4
    cost_matrix = [[400, -200, -300, -400, -500],   # 0
                   [-100, 800, -300, -400, -500],   # 1
                   [-100, -200, 1200, -400, -500],   # 2  TRUE VALUES
                   [-100, -200, -300, 1600, -500],   # 3
                   [-100, -200, -300, -400, 2000]]   # 4
    
    total = 0
    for r in range(0, 5):
        for c in range(0, 5):
            total = total + cost_matrix[r][c] * conf_matrix[r][c]

# # OR... THIS WORKS      total = np.dot(np.array(conf_matrix).ravel(), np.array(cost_matrix).ravel())
#     return total

***
# Section: 2 - Pre-Processing of the Dataset
- Use the provided dataset.  Split it into a Training dataset and a Test dataset based on the class attribute.  Keep them separate and use the Training dataset for training/tuning and the Test dataset for testing. For consistency, use the **train_test_split** operation available in SciKit Learn (use a specific random seed, so it is reproducible).
  - from sklearn.model_selection import train_test_split
  - X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
***

***
## Section: 2.1 - Explore the attributes
- As in Homework 1, explore the attributes briefly. Reference the website listed in the Introduction.
- Provide basic statistics for the attributes.
- List which attributes are Nominal (even though they are encoded as numbers), Ordinal, Interval, Ratio.
***

* id: patient's identifier
* age: patient's age
* sex: 1 is male, 0 is female
* cp: 9 cp: chest pain type; Value 1: typical angina;Value 2: atypical angina;Value 3: non-anginal pain;Value 4: asymptomatic
* trestbps: resting blood pressure, less than 120 is good better not more than 130
* chol: serum cholestoral in mg/dl
* fbs: fasting blood sugar, 1 if > 120 mg/dl (which is no good)
* restecg: resting electrocardiographic results;Value 0: normal；Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV);Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria
* thalach: maximum heart rate achieved, too high can indicate problem
* exang: exercise induced angina (1 = yes; 0 = no)
* oldpeak: ST depression induced by exercise relative to rest
* slope: the slope of the peak exercise ST segment;Value 1: upsloping;Value 2: flat;Value 3: downsloping
* ca: number of major vessels (0-3) colored by flourosopy
* thal: 3 = normal; 6 = fixed defect; 7 = reversable defect
* num: diagnosics

### Data Type:
* Nominal: id, age, sex, cp, restecg, fbs, slope, thal
* Ordinal:  exang,
* Interval: trespbps, chol, thalach, ca 
* Ratio: oldpeak

In [9]:
data = pd.read_csv("data2.csv")
# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
X_train, X_test, y_train, y_test = train_test_split(data.drop('num',axis=1), 
                                                    data['num'], test_size=0.33, random_state=42)
data

id  age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0      1   63    1   1       145   233    1        2      150      0      2.3   
1      2   67    1   4       160   286    0        2      108      1      1.5   
2      3   67    1   4       120   229    0        2      129      1      2.6   
3      4   37    1   3       130   250    0        0      187      0      3.5   
4      5   41    0   2       130   204    0        2      172      0      1.4   
..   ...  ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
277  278   39    0   3       138   220    0        0      152      0      0.0   
278  279   57    1   2       154   232    0        2      164      0      0.0   
279  280   58    0   4       130   197    0        0      131      0      0.6   
280  281   57    1   4       110   335    0        0      143      1      3.0   
281  282   47    1   3       130   253    0        0      179      0      0.0   

     slope  ca  thal  num  
0        3   0     6    0  
1        2   3     3    2  
2        2   2     7    1  
3        3   0     3    0  
4        1   0     3    0  
..     ...  ..   ...  ...  
277      2   0     3    0  
278      1   1     3    1  
279      2   0     3    0  
280      2   1     7    2  
281      1   0     3    0  

[282 rows x 15 columns]

In [10]:
data.describe()

id         age         sex          cp    trestbps        chol  \
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean   141.500000   54.411348    0.677305    3.163121  131.563830  249.092199   
std     81.550598    9.053083    0.468338    0.955405   17.757496   51.217546   
min      1.000000   29.000000    0.000000    1.000000   94.000000  126.000000   
25%     71.250000   48.000000    0.000000    3.000000  120.000000  213.000000   
50%    141.500000   55.000000    1.000000    3.000000  130.000000  244.000000   
75%    211.750000   61.000000    1.000000    4.000000  140.000000  277.000000   
max    282.000000   77.000000    1.000000    4.000000  200.000000  564.000000   

              fbs     restecg     thalach       exang     oldpeak       slope  \
count  282.000000  282.000000  282.000000  282.000000  282.000000  282.000000   
mean     0.148936    1.014184  149.765957    0.326241    1.026950    1.585106   
std      0.356658    0.998118   22.923869    0.469670    1.138825    0.609700   
min      0.000000    0.000000   71.000000    0.000000    0.000000    1.000000   
25%      0.000000    0.000000  133.250000    0.000000    0.000000    1.000000   
50%      0.000000    2.000000  153.500000    0.000000    0.800000    2.000000   
75%      0.000000    2.000000  165.750000    1.000000    1.600000    2.000000   
max      1.000000    2.000000  202.000000    1.000000    6.200000    3.000000   

               ca        thal         num  
count  282.000000  282.000000  282.000000  
mean     0.595745    4.581560    0.907801  
std      1.236910    2.248467    1.224894  
min     -9.000000   -9.000000    0.000000  
25%      0.000000    3.000000    0.000000  
50%      0.000000    3.000000    0.000000  
75%      1.000000    7.000000    2.000000  
max      3.000000    7.000000    4.000000

***
## Section: 2.2 - Revise the dataset
- Review the meanings of the attributes and consider removing redundant or (likely) irrelevant attributes, combining attributes, etc., to reduce the number of attributes.
- (You may choose to use techniques such as those you used in Homework 1 to analyze the impacts of individual attributes on the CLASS attribute, but you need not do a “deep” analysis.)
- Describe what you chose to do (and not do), and why.
***

***
## Section: 2.3 - Transform the attributes
- Consider transforming the remaining attributes (e.g., using the data dictionary to replace the numbers with text values for some attributes – this might or might not be useful), normalizing / scaling values, encoding labels (if necessary), etc.
- Describe what you chose to do (and not do), and why.
***

***
# Section: 3 - Evaluation of the Off-The-Shelf KNN Classifier
- Select the KNN classifier from the SciKit Learn library and run it on the dataset.
***

***
## Section: 3.1 - Configure the off-the-shelf KNN classifier
- Use the KNeighborsClassifier from the SciKit Learn library
- Explain all setup, parameters and execution options you chose to set, and why.
***

***
## Section: 3.2 - Run and evaluate the classifier
- Try several values of the K parameter and compare the results.
- Evaluate the performance of the classifier, using the evaluation method you defined above.
***

***
## Section: 3.3 - Evaluate the choice of the KNN classifier
- What characteristics of the problem and data made KNN a good or bad choice?
***

***
# Section: 4 - Evaluation of Off-The-Shelf Classifier #2
- As with the KNN classifier above, choose another classifier from the SciKit Learn library (Decision Tree, SVM, Logistic Regression, etc.) and run it on the dataset.
***

***
## Section: 4.1 - Configure the classifier
- Use the appropriate classifier from the SciKit Learn library.
- Explain all setup, parameters and execution options you chose to set, and why.
***

***
## Section: 4.2 - Run and evaluate the classifier
- Try several values of the parameters (if appropriate) and compare the results.
- Evaluate the performance of the classifier, using the evaluation method you defined above.
***

***
## Section: 4.3 - Evaluate the choice of the classifier
- What characteristics of the problem and data made the classifier a good or bad choice?
***

***
# Section: 5 - Evaluation of Off-The-Shelf Classifier #3
- As with the KNN classifier above, choose another classifier from the SciKit Learn library (Decision Tree, SVM, Logistic Regression, etc.) and run it on the dataset.
***

***
## Section: 5.1 - Configure the classifier
- Use the appropriate classifier from the SciKit Learn library.
- Explain all setup, parameters and execution options you chose to set, and why.
***

***
## Section: 5.2 - Run and evaluate the classifier
- Try several values of the parameters (if appropriate) and compare the results.
- Evaluate the performance of the classifier, using the evaluation method you defined above.
***

***
## Section: 5.3 - Evaluate the choice of the classifier
- What characteristics of the problem and data made the classifier a good or bad choice?
***

***
# Section: 6 - Comparison of the Three Classifiers
***

***
## Section: 6.1 - Compare the performance of these classifiers to each other
- What are their strong and weak points?
***

***
## Section: 6.2 - Choose a Best Classifier
- Choose one of the three classifiers as best and explain why.
***

***
# Section: 7 - Conclusions
- Write a paragraph on what you discovered or learned from this homework.
***

***
### END-OF-SUBMISSION
***